File: swift/Sources/Wakewords/Preprocessing.swift
import Foundation

public enum Preprocessing {
    public static let modelSampleRate = 16_000
    static let nFFT = 512
    static let winLength = 400
    static let hopLength = 160
    public static let nMels = 64
    public static let nMFCC = 64
    static let logMelEpsilon: Float = 1e-6

    public struct MFCCFeatures: Sendable {
        public let data: [Float]
        public let features: Int
        public let frames: Int

        public init(data: [Float], features: Int, frames: Int) {
            self.data = data
            self.features = features
            self.frames = frames
        }
    }

    public static func mfccFeatures(samples: [Float]) -> MFCCFeatures {
        let cache = MFCCHelper.cache
        let padded = reflectPad(samples: samples, pad: nFFT / 2)
        let frames = ((padded.count - nFFT) / hopLength) + 1
        var mfcc = Array(repeating: Float.zero, count: nMFCC * frames)
        var spectrum = Array(repeating: Float.zero, count: nFFT / 2 + 1)
        var mel = Array(repeating: Float.zero, count: nMels)
        var frameSamples = Array(repeating: Float.zero, count: nFFT)

        for frame in 0..<frames {
            let offset = frame * hopLength
            for index in 0..<nFFT {
                frameSamples[index] = padded[offset + index] * cache.window[index]
            }

            for bin in 0...(nFFT / 2) {
                var real: Float = 0
                var imag: Float = 0
                let cosRow = cache.cos[bin]
                let sinRow = cache.sin[bin]
                for index in 0..<nFFT {
                    let value = frameSamples[index]
                    real += value * cosRow[index]
                    imag -= value * sinRow[index]
                }
                spectrum[bin] = real * real + imag * imag
            }

            for melIndex in 0..<nMels {
                var energy: Float = 0
                for filter in cache.melFilters[melIndex] {
                    energy += spectrum[filter.bin] * filter.weight
                }
                mel[melIndex] = log(energy + logMelEpsilon)
            }

            for coeff in 0..<nMFCC {
                var value: Float = 0
                for melIndex in 0..<nMels {
                    value += mel[melIndex] * cache.dct[melIndex][coeff]
                }
                mfcc[(coeff * frames) + frame] = value
            }
        }

        return MFCCFeatures(data: mfcc, features: nMFCC, frames: frames)
    }

    public static func resample(samples: [Float], sourceRate: Int, targetRate: Int = modelSampleRate) -> [Float] {
        guard sourceRate != targetRate else { return samples }
        let targetLength = max(1, Int(round(Double(samples.count * targetRate) / Double(sourceRate))))
        var result = Array(repeating: Float.zero, count: targetLength)
        let scale = Float(samples.count - 1) / Float(max(1, targetLength - 1))

        for index in 0..<targetLength {
            let position = Float(index) * scale
            let left = Int(floor(position))
            let right = min(left + 1, samples.count - 1)
            let weight = position - Float(left)
            result[index] = (samples[left] * (1 - weight)) + (samples[right] * weight)
        }

        return result
    }
}

private enum MFCCHelper {
    static let cache = Cache(
        window: paddedHannWindow(),
        cos: dftTable(using: Foundation.cos),
        sin: dftTable(using: Foundation.sin),
        melFilters: melFilterbank(),
        dct: dctMatrix()
    )

    struct Filter: Sendable {
        let bin: Int
        let weight: Float
    }

    struct Cache: Sendable {
        let window: [Float]
        let cos: [[Float]]
        let sin: [[Float]]
        let melFilters: [[Filter]]
        let dct: [[Float]]
    }

    static func paddedHannWindow() -> [Float] {
        var window = Array(repeating: Float.zero, count: Preprocessing.nFFT)
        let start = (Preprocessing.nFFT - Preprocessing.winLength) / 2
        for index in 0..<Preprocessing.winLength {
            let value = 0.5 - (0.5 * Foundation.cos((2 * Double.pi * Double(index)) / Double(Preprocessing.winLength)))
            window[start + index] = Float(value)
        }
        return window
    }

    static func dftTable(using function: (Double) -> Double) -> [[Float]] {
        (0...(Preprocessing.nFFT / 2)).map { bin in
            (0..<Preprocessing.nFFT).map { index in
                Float(function((2 * Double.pi * Double(bin * index)) / Double(Preprocessing.nFFT)))
            }
        }
    }

    static func melFilterbank() -> [[Filter]] {
        let melMin = hzToMel(0)
        let melMax = hzToMel(Float(Preprocessing.modelSampleRate) / 2)
        let points = (0..<(Preprocessing.nMels + 2)).map { index in
            let mel = melMin + ((melMax - melMin) * Float(index) / Float(Preprocessing.nMels + 1))
            return melToHz(mel)
        }
        let fftFreqs = (0...(Preprocessing.nFFT / 2)).map { bin in
            Float(bin * Preprocessing.modelSampleRate) / Float(Preprocessing.nFFT)
        }

        return (0..<Preprocessing.nMels).map { melIndex in
            let lower = points[melIndex]
            let center = points[melIndex + 1]
            let upper = points[melIndex + 2]

            return fftFreqs.enumerated().compactMap { bin, frequency in
                let weight: Float
                if frequency >= lower && frequency <= center {
                    weight = (frequency - lower) / (center - lower)
                } else if frequency > center && frequency <= upper {
                    weight = (upper - frequency) / (upper - center)
                } else {
                    weight = 0
                }
                return weight > 0 ? Filter(bin: bin, weight: weight) : nil
            }
        }
    }

    static func hzToMel(_ frequency: Float) -> Float {
        2595 * log10(1 + (frequency / 700))
    }

    static func melToHz(_ mel: Float) -> Float {
        700 * (pow(10, mel / 2595) - 1)
    }

    static func dctMatrix() -> [[Float]] {
        (0..<Preprocessing.nMels).map { mel in
            (0..<Preprocessing.nMFCC).map { coeff in
                var value = Foundation.cos((Double.pi / Double(Preprocessing.nMels)) * (Double(mel) + 0.5) * Double(coeff))
                if coeff == 0 {
                    value *= 1 / sqrt(2.0)
                }
                return Float(value * sqrt(2.0 / Double(Preprocessing.nMels)))
            }
        }
    }

    static func reflectPad(samples: [Float], pad: Int) -> [Float] {
        var result = Array(repeating: Float.zero, count: samples.count + (pad * 2))
        for index in 0..<pad {
            result[index] = samples[pad - index]
            result[pad + samples.count + index] = samples[samples.count - 2 - index]
        }
        result.replaceSubrange(pad..<(pad + samples.count), with: samples)
        return result
    }
}

private func paddedHannWindow() -> [Float] {
    MFCCHelper.paddedHannWindow()
}

private func dftTable(using function: (Double) -> Double) -> [[Float]] {
    MFCCHelper.dftTable(using: function)
}

private func melFilterbank() -> [[MFCCHelper.Filter]] {
    MFCCHelper.melFilterbank()
}

private func dctMatrix() -> [[Float]] {
    MFCCHelper.dctMatrix()
}

private func reflectPad(samples: [Float], pad: Int) -> [Float] {
    MFCCHelper.reflectPad(samples: samples, pad: pad)
}

File: js/src/preprocessing.js
const MODEL_SAMPLE_RATE = 16000;
const N_FFT = 512;
const WIN_LENGTH = 400;
const HOP_LENGTH = 160;
const N_MELS = 64;
const N_MFCC = 64;
const LOG_MEL_EPSILON = 1e-6;

let mfccCache = null;

export function mfccFeatures(samples) {
  const cache = getMfccCache();
  const padded = reflectPad(samples, N_FFT / 2);
  const frames = Math.floor((padded.length - N_FFT) / HOP_LENGTH) + 1;
  const mfcc = new Float32Array(N_MFCC * frames);
  const spectrum = new Array(N_FFT / 2 + 1).fill(0);
  const mel = new Array(N_MELS).fill(0);
  const frameSamples = new Array(N_FFT).fill(0);

  for (let frame = 0; frame < frames; frame += 1) {
    const offset = frame * HOP_LENGTH;
    for (let index = 0; index < N_FFT; index += 1) {
      frameSamples[index] = padded[offset + index] * cache.window[index];
    }
    for (let bin = 0; bin <= N_FFT / 2; bin += 1) {
      let real = 0;
      let imag = 0;
      const cosRow = cache.cos[bin];
      const sinRow = cache.sin[bin];
      for (let index = 0; index < N_FFT; index += 1) {
        const value = frameSamples[index];
        real += value * cosRow[index];
        imag -= value * sinRow[index];
      }
      spectrum[bin] = real * real + imag * imag;
    }

    for (let melIndex = 0; melIndex < N_MELS; melIndex += 1) {
      let energy = 0;
      const filter = cache.melFilters[melIndex];
      for (const [bin, weight] of filter) {
        energy += spectrum[bin] * weight;
      }
      mel[melIndex] = Math.log(energy + LOG_MEL_EPSILON);
    }

    for (let coeff = 0; coeff < N_MFCC; coeff += 1) {
      let value = 0;
      for (let melIndex = 0; melIndex < N_MELS; melIndex += 1) {
        value += mel[melIndex] * cache.dct[melIndex][coeff];
      }
      mfcc[coeff * frames + frame] = value;
    }
  }

  return { data: mfcc, features: N_MFCC, frames };
}

export function resample(samples, sourceRate, targetRate = MODEL_SAMPLE_RATE) {
  if (sourceRate === targetRate) return new Float32Array(samples);
  const targetLength = Math.max(1, Math.round((samples.length * targetRate) / sourceRate));
  const result = new Float32Array(targetLength);
  const scale = (samples.length - 1) / Math.max(1, targetLength - 1);
  for (let index = 0; index < targetLength; index += 1) {
    const position = index * scale;
    const left = Math.floor(position);
    const right = Math.min(left + 1, samples.length - 1);
    const weight = position - left;
    result[index] = samples[left] * (1 - weight) + samples[right] * weight;
  }
  return result;
}

function getMfccCache() {
  if (mfccCache) return mfccCache;
  mfccCache = {
    window: paddedHannWindow(),
    cos: dftTable(Math.cos),
    sin: dftTable(Math.sin),
    melFilters: melFilterbank(),
    dct: dctMatrix(),
  };
  return mfccCache;
}

function paddedHannWindow() {
  const window = new Float32Array(N_FFT);
  const start = Math.floor((N_FFT - WIN_LENGTH) / 2);
  for (let index = 0; index < WIN_LENGTH; index += 1) {
    window[start + index] = 0.5 - 0.5 * Math.cos((2 * Math.PI * index) / WIN_LENGTH);
  }
  return window;
}

function dftTable(fn) {
  const rows = [];
  for (let bin = 0; bin <= N_FFT / 2; bin += 1) {
    const row = new Float32Array(N_FFT);
    for (let index = 0; index < N_FFT; index += 1) {
      row[index] = fn((2 * Math.PI * bin * index) / N_FFT);
    }
    rows.push(row);
  }
  return rows;
}

function melFilterbank() {
  const filters = Array.from({ length: N_MELS }, () => []);
  const fMin = 0;
  const fMax = MODEL_SAMPLE_RATE / 2;
  const melMin = hzToMel(fMin);
  const melMax = hzToMel(fMax);
  const points = [];
  for (let index = 0; index < N_MELS + 2; index += 1) {
    const mel = melMin + ((melMax - melMin) * index) / (N_MELS + 1);
    points.push(melToHz(mel));
  }
  const fftFreqs = [];
  for (let bin = 0; bin <= N_FFT / 2; bin += 1) {
    fftFreqs.push((bin * MODEL_SAMPLE_RATE) / N_FFT);
  }
  for (let melIndex = 0; melIndex < N_MELS; melIndex += 1) {
    const lower = points[melIndex];
    const center = points[melIndex + 1];
    const upper = points[melIndex + 2];
    for (let bin = 0; bin < fftFreqs.length; bin += 1) {
      const freq = fftFreqs[bin];
      let weight = 0;
      if (freq >= lower && freq <= center) {
        weight = (freq - lower) / (center - lower);
      } else if (freq > center && freq <= upper) {
        weight = (upper - freq) / (upper - center);
      }
      if (weight > 0) filters[melIndex].push([bin, weight]);
    }
  }
  return filters;
}

function hzToMel(freq) {
  return 2595 * Math.log10(1 + freq / 700);
}

function melToHz(mel) {
  return 700 * (10 ** (mel / 2595) - 1);
}

function dctMatrix() {
  const matrix = [];
  for (let mel = 0; mel < N_MELS; mel += 1) {
    const row = new Float32Array(N_MFCC);
    for (let coeff = 0; coeff < N_MFCC; coeff += 1) {
      let value = Math.cos((Math.PI / N_MELS) * (mel + 0.5) * coeff);
      if (coeff === 0) value *= 1 / Math.sqrt(2);
      row[coeff] = value * Math.sqrt(2 / N_MELS);
    }
    matrix.push(row);
  }
  return matrix;
}

function reflectPad(samples, pad) {
  const result = new Float32Array(samples.length + pad * 2);
  for (let index = 0; index < pad; index += 1) {
    result[index] = samples[pad - index];
    result[pad + samples.length + index] = samples[samples.length - 2 - index];
  }
  result.set(samples, pad);
  return result;
}
